### PR TITLE
Fix e2e training edge cases

### PR DESCRIPTION
**Pull request by joejuzl**
**Proposed changes**:
- Fixes are variety of training scenarios outlined in #7129 
- fixes https://github.com/RasaHQ/rasa/issues/7360

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by wochinge** [comment]
@joejuzl I'm offline tomorrow and won't get to it today. Is next week fine? Otherwise it would be great to request review from someone else.

**Comment by joejuzl** [comment]
@wochinge I've requested review from another recommended backender. If it's still open next week you can have a look too 😄

**Comment by degiz** [comment]
I'll review tomorrow morning 👍

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 1]
```diff
@@ -266,7 +266,7 @@ def _parse_step(self, step: Union[Text, Dict[Text, Any]]) -> None:
             )
         elif KEY_USER_MESSAGE in step.keys():
             self._parse_user_message(step)
-        elif KEY_USER_INTENT in step.keys() or KEY_USER_MESSAGE in step.keys():
+        elif KEY_USER_INTENT in step.keys():
             self._parse_labeled_user_utterance(step)
```
Comment (Ghostvv) [inline rasa/shared/core/training_data/story_reader/yaml_story_reader.py:270]: are you sure you don't break e2e tests?

File: rasa/model.py [hunk 2]
```diff
@@ -451,6 +454,9 @@ def should_retrain(
             ),
         )
 
+        if has_e2e_examples and fingerprint_comparison.should_retrain_nlu():
```
Comment (joejuzl) [inline rasa/model.py:457]: We should retrain core if nlu data changes and there are e2e stories.

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py (hunk 1 above)
Comment (joejuzl) [reply rasa/shared/core/training_data/story_reader/yaml_story_reader.py:270]: The removed code is identical to the above `elif` statement, so totally redundant.

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 3]
```diff
@@ -317,9 +317,16 @@ def _parse_raw_user_utterance(self, step: Dict[Text, Any]) -> Optional[UserUtter
         return UserUttered(None, intent, entities)
 
     def _parse_user_message(self, step: Dict[Text, Any]) -> None:
-        is_end_to_end_utterance = KEY_USER_INTENT not in step
+        has_user_text_and_intent = KEY_USER_INTENT in step
 
-        if is_end_to_end_utterance:
+        if has_user_text_and_intent and not self.use_e2e:
```
[the remark this answers is not in the document]
Comment (joejuzl) [reply rasa/shared/core/training_data/story_reader/yaml_story_reader.py:322]: Just so we are clear `self.use_e2e` indicates an e2e conversation _test_ story right?

This change was to address the point:

> do not allow both intent: and user: in training data

And without it I think we get some tensor-flow input shape errors when training core. (If we have mixed `user` and `intent` steps in the training story)

File: rasa/shared/importers/importer.py [hunk 4]
```diff
@@ -100,8 +100,7 @@ def load_core_importer_from_config(
         importer = TrainingDataImporter.load_from_config(
             config_path, domain_path, training_data_paths, TrainingType.CORE
         )
-
-        return CoreDataImporter(importer)
+        return importer
```
Comment (joejuzl) [inline rasa/shared/importers/importer.py:103]: This allows us to check for e2e data in the NLU training data when using the `rasa core` command.

File: rasa/train.py [hunk 5]
```diff
@@ -174,7 +174,7 @@ async def _train_async_internal(
             additional_arguments=nlu_additional_arguments,
         )
 
-    if nlu_data.can_train_nlu_model():
+    if nlu_data.is_empty():
```
Comment (joejuzl) [inline rasa/train.py:177]: This means we will train NLU even if the only NLU data is from e2e stories.

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/shared/core/training_data/story_reader/yaml_story_reader.py:322]: yes, but we should fix it inside core featurizers, I can quickly do it and push to your branch

[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/shared/core/training_data/story_reader/yaml_story_reader.py:322]: pushed a fix, let me know if it works, when you allow both intent and text go through

File: tests/conftest.py [hunk 6]
```diff
@@ -191,6 +205,22 @@ async def trained_rasa_model(
     return trained_stack_model_path
 
 
+@pytest.fixture(scope="session")
+async def trained_simple_rasa_model(
```
Comment (degiz) [inline tests/conftest.py:209]: Does it bring some advantages over existing `trained_rasa_model()`?

File: tests/test_train.py [hunk 7]
```diff
@@ -358,3 +359,227 @@ def test_train_nlu_autoconfig(
...
+
+def mock_nlu_training(monkeypatch):
+    return mock_async(monkeypatch, rasa.nlu, rasa.nlu.train.__name__)
+
+
+def new_model_path_in_same_dir(old_model_path) -> str:
+    return str(Path(old_model_path).parent / (secrets.token_hex(8) + ".tar.gz"))
+
+
+class TestE2e:
```
Comment (degiz) [inline tests/test_train.py:386]: Maybe makes sense to extract `e2e` to a separate file?

File: rasa/shared/importers/importer.py (hunk 4 above)
Comment (joejuzl) [reply rasa/shared/importers/importer.py:103]: The `get_stories` in `CoreDataImporter` just calls the `get_stories` of the importer it wraps (the one we are returning directly here). The only real difference with `CoreDataImporter` is that `get_nlu_data` is overridden to return an empty `TrainingData`.

File: rasa/shared/nlu/training_data/message.py [hunk 8]
```diff
@@ -324,13 +324,14 @@ def _combine_features(
 
         return combined_features
 
-    def is_core_message(self) -> bool:
-        """Checks whether the message is a core message or not.
+    def is_core_or_domain_message(self) -> bool:
+        """Checks whether the message is a core message or from the domain.
```
[the remark this answers is not in the document]
Comment (joejuzl) [reply rasa/shared/nlu/training_data/message.py:328]: hmm maybe "from the domain" it not clear (or even correct). The other kind of messages are actions. So default actions, or actions used in stories.

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 9]
```diff
@@ -319,9 +319,9 @@ def _parse_labeled_user_utterance(self, step: Dict[Text, Any]) -> None:
             self.current_step_builder.add_user_messages([utterance])
 
     def _parse_user_message(self, step: Dict[Text, Any]) -> None:
-        is_end_to_end_utterance = KEY_USER_INTENT not in step
+        has_user_text_and_intent = KEY_USER_INTENT in step
```
[the remark this answers is not in the document]
Comment (joejuzl) [reply rasa/shared/core/training_data/story_reader/yaml_story_reader.py:322]: It was not initially obvious to me as you have to read the big if else statement that calls this method. But I agree it is a bit wordy. Maybe: `also_has_intent`?

File: rasa/shared/nlu/training_data/training_data.py [hunk 10]
```diff
@@ -668,7 +676,7 @@ def is_empty(self) -> bool:
         ]
         return not any([len(lst) > 0 for lst in lists_to_check])
 
-    def can_train_nlu_model(self) -> bool:
+    def contains_no_pure_nlu_data(self) -> bool:
```
[the remark this answers is not in the document]
Comment (joejuzl) [reply rasa/shared/nlu/training_data/training_data.py:679]: Haha yeah I couldn't think of a good name... Open to suggestions, but it would have to be a negation of `contains_any_nlu_data`.

File: rasa/train.py [hunk 11]
```diff
@@ -349,7 +349,19 @@ async def train_core_async(
     file_importer = TrainingDataImporter.load_core_importer_from_config(
         config, domain, [stories]
     )
-    domain = await file_importer.get_domain()
+    stories, nlu_data, domain = await asyncio.gather(
```
[the remark this answers is not in the document]
Comment (joejuzl) [reply rasa/train.py:352]: I can use it below rather than reloading.

File: tests/conftest.py (hunk 6 above)
Comment (joejuzl) [reply tests/conftest.py:209]: To be honest, `trained_rasa_model()` is not working on `e2e` branch right now. Something to do with entities. And I needed a way to test this.

File: tests/test_train.py (hunk 7 above)
Comment (joejuzl) [reply tests/test_train.py:386]: It means losing the module name to test file name mapping, but if you think this is too big/different I can move it.

File: rasa/core/featurizers/tracker_featurizers.py [hunk 12]
```diff
@@ -106,12 +106,22 @@ def _create_entity_tags(
 
     @staticmethod
     def _entity_data(event: UserUttered) -> Dict[Text, Any]:
-        if event.text:
+        # train stories support both text and intent,
```
[the remark this answers is not in the document]
Comment (joejuzl) [reply rasa/core/featurizers/tracker_featurizers.py:109]: @Ghostvv maybe one for you

File: tests/test_train.py (hunk 7 above)
Comment (joejuzl) [reply tests/test_train.py:386]: I don't feel strongly either way.

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py (hunk 1 above)
Comment (wochinge) [reply rasa/shared/core/training_data/story_reader/yaml_story_reader.py:281]: currently fixing this stuff in https://github.com/RasaHQ/rasa/pull/7388/files as well

File: rasa/core/featurizers/tracker_featurizers.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/core/featurizers/tracker_featurizers.py:109]: if you want, sure, what do you propose?

File: tests/conftest.py (hunk 6 above)
Comment (joejuzl) [reply tests/conftest.py:209]: Currently we can not use `trained_rasa_model` here as there are a couple of different issues with how we fingerprint stories at the moment using `as_story_string()`.

These will be fixed in a separate PR.
So I think best for now to use this new `trained_simple_rasa_model`, and then unify it when we fix those bugs.

**Review by degiz — APPROVED** [review]
Thanks for addressing the comments 👍
